src/mapreduce/worker.go

```diff
@@ -128,3 +141,63 @@ func RunWorker(MasterAddress string, me string,
 	wk.l.Close()
 	debug("RunWorker %s exit\n", me)
 }
+
+// The function servers as an intermediate check so that
+// after the shutdown function is called the listener will
+// be closed.
+func (wk *Worker) ServeAndCheck(rpcs *rpc.Server, conn net.Conn) {
+	rpcs.ServeConn(conn)
+	wk.Lock()
+	wk.wg.Done()
+	if wk.done == true {
```

Comment:
boxiao, could you think about another way to exit worker's server, Not make goroutine to close worker's server?